### ISSUE TITLE
Autoconversion to Tensor in functions

### ISSUE THREAD

**Issue by mayou36** [tensorflow/tensorflow#30029]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 19.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary, pip
- TensorFlow version (use command below):1.14.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
calling certain functions with non-Tensors (but convertible to) fails, since the object is not converted to a tensor. The function actually only expect Tensors.

e.g. `tf.math.real(1.0)` fails because `1.` has no attribute dtype (which it will only have after the conversion to a tensor).

**Describe the expected behavior**
(My understanding goes that:) functions that take Tensors, such as `tf.math.real`, can also take anything that can be converted to a Tensor, such as Python floats or objects with a registered conversion function. Namely,
```
tf.math.real(tf.convert_to_tensor(something))
tf.math.real(something)
```
are _equivalent_.

I would expect `tf.math.real(1.)` to return the real part of the tensor `tf.convert_to_tensor(1.)`.

This causes a big problem with the (actual beautiful!) registration of conversion functions. The code contains only a minimal example.

Am I mistaken with the expected behavior? And if so, this may could be made clear in the docs, that _only_ Tensors are taken vs Tensor-like objects. E.g. the docs of `tf.math.round` and `tf.math.real` leave no clue in which to use tensors and in which Tensor-like objects.

**Code to reproduce the issue**
short version
```
real_python = tf.math.real(1.)  # <- fails
```
but of course also fails for any custom defined Tensor-like object
```
from tensorflow.python import ops

class MyTensor():
    def _dense_var_to_tensor(self, dtype, name, as_ref):
        return tf.constant(42, dtype=dtype)


def _dense_var_to_tensor(var, dtype=None, name=None, as_ref=False):
    return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)


ops.register_tensor_conversion_function(MyTensor, _dense_var_to_tensor)

my_t1 = MyTensor()
t1 = tf.convert_to_tensor(my_t1)  # <- works
square = tf.math.round(my_t1)  # <- works
real_python = tf.math.real(1.)  # <- fails
real = tf.math.real(my_t1)  # <- fails
```

**Comment by yongtang** [comment]
Added a PR #30049 for the fix.
